### ISSUE TITLE
SSM connection randomly fails with "Failed to create temporary directory" because "File exists"

### ISSUE THREAD

**Issue by fab-10** [ansible-collections/community.aws#1582]
### Summary

When using `aws_ssm` connection plugin, I get this error:

```
fatal: [besu-2]: UNREACHABLE! => {"changed": false, "msg": "Failed to create temporary directory. In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\", for more error information use -vvv. Failed command was: ( umask 77 && mkdir -p \"` echo /tmp/.ansible-ssm `\"&& mkdir \"` echo /tmp/.ansible-ssm/ansible-tmp-1667381646.1546528-616-871393679439 `\" && echo ansible-tmp-1667381646.1546528-616-871393679439=\"` echo /tmp/.ansible-ssm/ansible-tmp-1667381646.1546528-616-871393679439 `\" ), exited with result 1, stdout output: mkdir: cannot create directory ‘/tmp/.ansible-ssm/ansible-tmp-1667381646.1546528-616-871393679439’: File exists\r\r", "unreachable": true}
```

it happens at random and it is transient, so rerunning could work or fail at another task.
I have tried to change the tmp path, but it does not help, the error says that the directory already exists, so I am wondering if it the result of a retry.
Could a solution check if the directory already exists and delete/reuse it?

### Issue Type

Bug Report

### Component Name

aws_ssm

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.13.3]
  config file = /ansible/ansible.cfg
  configured module search path = ['/ansible/library']
  ansible python module location = /usr/local/lib/python3.10/dist-packages/ansible
  ansible collection location = /home/ansible/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.10.6 (main, Aug 10 2022, 11:40:04) [GCC 11.3.0]
  jinja version = 3.1.2
  libyaml = True
```

### Collection Versions

```console (paste below)
$ ansible-galaxy collection list
Collection        Version
----------------- -------
amazon.aws        4.0.0
ansible.posix     1.4.0
community.aws     4.0.0
community.docker  3.0.1
community.general 5.4.0
```

### AWS SDK versions

```console (paste below)
$ pip show boto boto3 botocore
WARNING: Package(s) not found: boto
Name: boto3
Version: 1.24.52
Summary: The AWS SDK for Python
Home-page: https://github.com/boto/boto3
Author: Amazon Web Services
Author-email:
License: Apache License 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: botocore, jmespath, s3transfer
Required-by:
---
Name: botocore
Version: 1.27.96
Summary: Low-level, data-driven core of boto 3.
Home-page: https://github.com/boto/botocore
Author: Amazon Web Services
Author-email:
License: Apache License 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: jmespath, python-dateutil, urllib3
Required-by: boto3, s3transfer
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed
DEFAULT_FILTER_PLUGIN_PATH(/ansible/ansible.cfg) = ['/ansible/filter_plugins']
DEFAULT_FORKS(/ansible/ansible.cfg) = 1
DEFAULT_MODULE_PATH(/ansible/ansible.cfg) = ['/ansible/library']
DEFAULT_ROLES_PATH(/ansible/ansible.cfg) = ['/usr/share/ansible/roles', '/ansible/roles']
HOST_KEY_CHECKING(/ansible/ansible.cfg) = True
RETRY_FILES_ENABLED(/ansible/ansible.cfg) = False
USE_PERSISTENT_CONNECTIONS(/ansible/ansible.cfg) = True
```

### OS / Environment

Ubuntu 22.04

### Steps to Reproduce

```yaml (paste below)

```

### Expected Results

Task is executed

### Actual Results

```console (paste below)
fatal: [besu-2]: UNREACHABLE! => {"changed": false, "msg": "Failed to create temporary directory. In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\", for more error information use -vvv. Failed command was: ( umask 77 && mkdir -p \"` echo /tmp/.ansible-ssm `\"&& mkdir \"` echo /tmp/.ansible-ssm/ansible-tmp-1667381646.1546528-616-871393679439 `\" && echo ansible-tmp-1667381646.1546528-616-871393679439=\"` echo /tmp/.ansible-ssm/ansible-tmp-1667381646.1546528-616-871393679439 `\" ), exited with result 1, stdout output: mkdir: cannot create directory ‘/tmp/.ansible-ssm/ansible-tmp-1667381646.1546528-616-871393679439’: File exists\r\r", "unreachable": true}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibullbot** [comment]
Files identified in the description:
* [`plugins/connection/aws_ssm.py`](https://github.com/['ansible-collections/amazon.aws', 'ansible-collections/community.aws', 'ansible-collections/community.vmware']/blob/main/plugins/connection/aws_ssm.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by schniber** [comment]
Hello community,

Are there any updates on this issue ? I have been facing this one too with no clear way around.
=> For me this is one of the flagship features allowing to run ansible over ssm rather than over ssh.

Thanks again.

Bests.

**Comment by fab-10** [comment]
@schniber I have a workaround for the moment, that requires patching ansible core is your are willing to try it:

```
--- a/lib/ansible/plugins/shell/__init__.py
+++ b/lib/ansible/plugins/shell/__init__.py
@@ -169,7 +169,7 @@ class ShellBase(AnsiblePlugin):

         # use mkdir -p to ensure parents exist, but mkdir fullpath to ensure last one is created by us
         cmd = 'mkdir -p %s echo %s %s' % (self._SHELL_SUB_LEFT, basetmpdir, self._SHELL_SUB_RIGHT)
-        cmd += '%s mkdir %s echo %s %s' % (self._SHELL_AND, self._SHELL_SUB_LEFT, basetmp, self._SHELL_SUB_RIGHT)
+        cmd += '%s mkdir -p %s echo %s %s' % (self._SHELL_AND, self._SHELL_SUB_LEFT, basetmp, self._SHELL_SUB_RIGHT)
         cmd += ' %s echo %s=%s echo %s %s' % (self._SHELL_AND, basefile, self._SHELL_SUB_LEFT, basetmp, self._SHELL_SUB_RIGHT)

         # change the umask in a subshell to achieve the desired mode
```

**Comment by schniber** [comment]
Hello Fabio,

I applied your patch locally and this looks to be solving it. This a bug in the Ansible core itself rather than on this collection as you spotted.
I think that the bot which closed your initial issue here https://github.com/ansible/ansible/issues/79297 was mistaken.

Do you think it would make sense to re-open the issue in there and create a pull request with your patch ?

Thanks.

Bests.

**Comment by tremble** [comment]
@schniber @fab-10 the ansible/ansible issue wasn't closed "by mistake" as such, it was closed as a part of the split of Ansible Core vs Ansible Collections.  There were so many 'stale' issues that they decided to close them rather than try to migrate them when the code was moved to separate repos.

I've recently been digging into the aws_ssm plugin to try and clean up some of the issues.  Unfortunately I generally use the SSH connection plugin rather than the SSM connection plugin limiting my "production" experience with the SSM plugin.

We've just released version 5.2.0 which includes some significant work on the plugin.  Could you try and reproduce the issue using release 5.2.0 and `ansible-playbook -vvvv`.  It would be helpful to have the trace from a failed command.  For what it's worth I suspect you're right about the 'retries' being the origin of the issue, the question is if we can tune how things are retried.

**Comment by schniber** [comment]
Hello @tremble ! 

Thanks for your inputs, though I concluded it was a mistake since the fix spotted by @fab-10 seemed to resolve the issue and since it was nothing related to the collection hence my statement.

Now, I do agree that is one of the most interesting features of the SSM plugin which allows not to leverage SSH. 

Like you, I tended to privilege SSH over SSM du to the instability but I wanted to give this a serious try and hopefully go productively with it.

I will test the new collection release and remove temporarily fabio's fix and will feedback in here.

**Comment by fab-10** [comment]
@schniber I do not think this is a bug in ansible core, because the same code works well with SSH, so I am more inclined to think that there are interactions between SSM and temp files

**Comment by schniber** [comment]
Hello @tremble and @fab-10 
Release 5.2.0 seems to have solved the instability. I did quite few tests and also removed the patch fabio suggested and it all works fine from my side.
Keep it up ! 

Thanks !

**Comment by tremble** [comment]
@fab-10 

https://galaxy.ansible.com/community/aws

As far as I can tell the collection's been successfully released and should be available through `ansible-galaxy`.  I did forget to update GitHub's "release" which has now been corrected.

**Comment by tremble** [comment]
@schniber that's excellent news.  I'll give @fab-10 a couple of days and if the issue isn't reproducible with 5.2.0 I'll close out this issue.

**Comment by fab-10** [comment]
unfortunately I am not able to do a proper test in the next few days, so feel free to proceed on this.

**Comment by hectoralicea** [comment]
I too am having this problem.  The task looks like the below and its running in a loop.  It keeps stoping after 2 or 3 EC2s in the loop.   I'm using 5.2.0 of "https://github.com/ansible-collections/community.aws".  Maybe i should bump up ansible core?

```
- name: drain-node | ensure all non-DaemonSet pods running on each host has been evicted from {{ item.tags.Name }}
  ansible.builtin.shell: |
    kubectl get pods --all-namespaces -o custom-columns=NAME:.metadata.name,CONTROLLER:.metadata.ownerReferences[].kind,NAMESPACE:.metadata.namespace,NODE:.status.hostIP | grep {{ hostvars[item.instance_id]['private_ip_address'] }} | grep -v DaemonSet || true
  args:
    chdir: /root
  register: kc_get_pod_results
  retries: 30
  delay: 2
# no_log: true
  ignore_errors: yes
  until: 
  - (kc_get_pod_results.stdout_lines | length) == 0
  -  kc_get_pod_results.failed == false
```
